File: examples/dubbing/node/dubbingUtils.ts
import * as dotenv from "dotenv";
import * as fs from "fs";
import * as path from "path";
import { ElevenLabs } from "elevenlabs"; // Assuming an ElevenLabs TypeScript SDK exists

// Load environment variables
dotenv.config();

// Retrieve the API key
const ELEVENLABS_API_KEY = process.env.ELEVENLABS_API_KEY;
if (!ELEVENLABS_API_KEY) {
  throw new Error(
    "ELEVENLABS_API_KEY environment variable not found. " +
      "Please set the API key in your environment variables."
  );
}

const client = new ElevenLabs({ apiKey: ELEVENLABS_API_KEY });

function downloadDubbedFile(
  dubbingId: string,
  languageCode: string
): Promise<string> {
  return new Promise((resolve, reject) => {
    const dirPath = path.join("data", dubbingId);
    fs.mkdirSync(dirPath, { recursive: true });

    const filePath = path.join(dirPath, `${languageCode}.mp4`);
    const fileStream = fs.createWriteStream(filePath);
    const dubbedFileStream = client.dubbing.getDubbedFile(
      dubbingId,
      languageCode
    );
    dubbedFileStream.on("data", (chunk: Buffer) => {
      fileStream.write(chunk);
    });
    dubbedFileStream.on("end", () => {
      fileStream.end();
      resolve(filePath);
    });
    dubbedFileStream.on("error", reject);
  });
}

async function waitForDubbingCompletion(dubbingId: string): Promise<boolean> {
  const MAX_ATTEMPTS = 120;
  const CHECK_INTERVAL = 10000; // In milliseconds

  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
    const metadata = await client.dubbing.getDubbingProjectMetadata(dubbingId);
    if (metadata.status === "dubbed") {
      return true;
    } else if (metadata.status === "dubbing") {
      console.log(
        "Dubbing in progress... Will check status again in",
        CHECK_INTERVAL / 1000,
        "seconds."
      );
      await new Promise(resolve => setTimeout(resolve, CHECK_INTERVAL));
    } else {
      console.log("Dubbing failed:", metadata.error_message);
      return false;
    }
  }

  console.log("Dubbing timed out");
  return false;
}

File: examples/dubbing/python/dubbing_utils.py
import os
import time

from dotenv import load_dotenv
from elevenlabs.client import ElevenLabs

# Load environment variables
load_dotenv()

# Retrieve the API key
ELEVENLABS_API_KEY = os.getenv("ELEVENLABS_API_KEY")
if not ELEVENLABS_API_KEY:
    raise ValueError(
        "ELEVENLABS_API_KEY environment variable not found. "
        "Please set the API key in your environment variables."
    )

client = ElevenLabs(api_key=ELEVENLABS_API_KEY)


def download_dubbed_file(dubbing_id: str, language_code: str) -> str:
    """
    Downloads the dubbed file for a given dubbing ID and language code.

    Args:
        dubbing_id: The ID of the dubbing project.
        language_code: The language code for the dubbing.

    Returns:
        The file path to the downloaded dubbed file.
    """
    dir_path = f"data/{dubbing_id}"
    os.makedirs(dir_path, exist_ok=True)

    file_path = f"{dir_path}/{language_code}.mp4"
    with open(file_path, "wb") as file:
        for chunk in client.dubbing.get_dubbed_file(dubbing_id, language_code):
            file.write(chunk)

    return file_path


def wait_for_dubbing_completion(dubbing_id: str) -> bool:
    """
    Waits for the dubbing process to complete by periodically checking the status.

    Args:
        dubbing_id (str): The dubbing project id.

    Returns:
        bool: True if the dubbing is successful, False otherwise.
    """
    MAX_ATTEMPTS = 120
    CHECK_INTERVAL = 10  # In seconds

    for _ in range(MAX_ATTEMPTS):
        metadata = client.dubbing.get_dubbing_project_metadata(dubbing_id)
        if metadata.status == "dubbed":
            return True
        elif metadata.status == "dubbing":
            print(
                "Dubbing in progress... Will check status again in",
                CHECK_INTERVAL,
                "seconds.",
            )
            time.sleep(CHECK_INTERVAL)
        else:
            print("Dubbing failed:", metadata.error_message)
            return False

    print("Dubbing timed out")
    return False
